File: packages/typescript/src/exporter/http.ts
/**
 * HTTPS batched exporter.
 *
 * Mirrors agent_capture.exporter.http. 5xx/network errors retry with
 * exponential backoff; 4xx drops with a loud AC403 safelog entry.
 * Never throws into the producer.
 */

import { ErrorCode, logError } from "../internal/safelog.js";
import type { AgentCaptureSpan } from "../schema/span.js";

import type { SpanExporter } from "./base.js";
import { defaultRetryPolicy, type RetryPolicy, withRetry } from "./retry.js";

export interface HTTPExporterOptions {
  endpoint: string;
  authToken?: string;
  batchSize?: number;
  batchMaxWaitMs?: number;
  timeoutMs?: number;
  retryPolicy?: RetryPolicy;
  /** Inject for tests. */
  fetch?: typeof fetch;
}

class PermanentHTTPError extends Error {
  constructor(public readonly status: number, public readonly body: string) {
    super(`HTTP ${status}: ${body}`);
  }
}

export class HTTPExporter implements SpanExporter {
  private buffer: AgentCaptureSpan[] = [];
  private firstEnqueueTime: number | null = null;
  private droppedCount = 0;
  private isShutdown = false;
  private timer: ReturnType<typeof setTimeout> | null = null;
  private inFlight: Promise<void> | null = null;
  private readonly fetchImpl: typeof fetch;
  private readonly batchSize: number;
  private readonly batchMaxWaitMs: number;
  private readonly timeoutMs: number;
  private readonly retryPolicy: RetryPolicy;

  constructor(private readonly options: HTTPExporterOptions) {
    this.fetchImpl = options.fetch ?? globalThis.fetch;
    this.batchSize = options.batchSize ?? 100;
    this.batchMaxWaitMs = options.batchMaxWaitMs ?? 1000;
    this.timeoutMs = options.timeoutMs ?? 10_000;
    this.retryPolicy = options.retryPolicy ?? defaultRetryPolicy;
  }

  get dropped(): number {
    return this.droppedCount;
  }

  export(span: AgentCaptureSpan): void {
    if (this.isShutdown) return;
    this.buffer.push(span);
    if (this.firstEnqueueTime === null) {
      this.firstEnqueueTime = Date.now();
    }
    if (this.buffer.length >= this.batchSize) {
      void this.flush();
    } else {
      this.ensureTimer();
    }
  }

  private ensureTimer(): void {
    if (this.timer !== null) return;
    this.timer = setTimeout(() => {
      this.timer = null;
      void this.flush();
    }, this.batchMaxWaitMs);
  }

  private async flush(): Promise<void> {
    if (this.timer !== null) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    if (this.buffer.length === 0) {
      this.firstEnqueueTime = null;
      return;
    }
    const batch = this.buffer;
    this.buffer = [];
    this.firstEnqueueTime = null;
    const send = this.sendWithRetry(batch);
    this.inFlight = send.finally(() => {
      if (this.inFlight === send) this.inFlight = null;
    });
    await send;
  }

  private async sendWithRetry(batch: AgentCaptureSpan[]): Promise<void> {
    let payload: string;
    try {
      payload = JSON.stringify({ spans: batch });
    } catch (exc) {
      logError(
        ErrorCode.AC402,
        `HTTPExporter: serialization failed; dropping ${batch.length} spans`,
        exc,
      );
      this.droppedCount += batch.length;
      return;
    }
    try {
      await withRetry(
        async () => {
          const headers: Record<string, string> = {
            "content-type": "application/json",
          };
          if (this.options.authToken !== undefined) {
            headers.authorization = `Bearer ${this.options.authToken}`;
          }
          const controller = new AbortController();
          const timeoutId = setTimeout(() => controller.abort(), this.timeoutMs);
          let resp: Response;
          try {
            resp = await this.fetchImpl(this.options.endpoint, {
              method: "POST",
              body: payload,
              headers,
              signal: controller.signal,
            });
          } finally {
            clearTimeout(timeoutId);
          }
          if (resp.status >= 400 && resp.status < 500) {
            const body = await resp.text().catch(() => "");
            throw new PermanentHTTPError(resp.status, body.slice(0, 512));
          }
          if (!resp.ok) {
            throw new Error(`HTTP ${resp.status}`);
          }
        },
        {
          policy: this.retryPolicy,
          retryable: (exc) => !(exc instanceof PermanentHTTPError),
        },
      );
    } catch (exc) {
      if (exc instanceof PermanentHTTPError) {
        logError(
          ErrorCode.AC403,
          `HTTPExporter: ledger rejected batch of ${batch.length} spans with HTTP ${exc.status}: ${exc.body}`,
        );
      } else {
        logError(
          ErrorCode.AC404,
          `HTTPExporter: dropping ${batch.length} spans after retries`,
          exc,
        );
      }
      this.droppedCount += batch.length;
    }
  }

  async shutdown(_timeoutMs: number = 5000): Promise<void> {
    if (this.isShutdown) return;
    this.isShutdown = true;
    if (this.timer !== null) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    await this.flush();
    await this.inFlight;
  }
}

File: packages/python/src/agent_capture/exporter/http.py
"""HTTPS exporter — ships batches over HTTP to the vendor-cloud ledger.

Producer side (the agent's hot path) calls :meth:`export` which appends
the span to an internal batch buffer and returns immediately. A
background flush thread drains the buffer when either:

- the buffer reaches ``batch_size``, OR
- ``batch_max_wait_s`` has elapsed since the last flush.

On send failure: exponential backoff via :mod:`.retry` for transient
errors (5xx, network); permanent errors (4xx, payload too large) drop
after one attempt with a loud safelog entry. The agent never sees the
failure.

Cardinal rule (architecture doc §9.2): the agent must always win. If
the ledger is down for hours, we drop oldest spans rather than queue
unboundedly. The :class:`BoundedQueueExporter` upstream owns that drop
policy; this exporter just *attempts* delivery and reports outcomes via
the drop counter.

Payload shape: a JSON object with ``{"spans": [<span_json>, ...]}``.
Each span is the Pydantic ``model_dump_json`` form. ``Content-Type:
application/json``. Authorization is optional and configured at
construction.
"""

from __future__ import annotations

import contextlib
import json
import threading
import time
from typing import Any

import httpx

from agent_capture._internal.safelog import ErrorCode, log_error
from agent_capture.exporter.base import SpanExporter
from agent_capture.exporter.retry import RetryPolicy, with_retry
from agent_capture.schema import Span


class HTTPExporter(SpanExporter):
    """Batched HTTPS POST to the vendor-cloud ledger.

    Args:
        endpoint: URL of the ledger ingest endpoint (HTTPS in production).
        auth_token: Optional bearer token sent as ``Authorization: Bearer``.
        batch_size: Flush as soon as the buffer holds this many spans.
        batch_max_wait_s: Flush at most this many seconds after the first
            span enters an empty buffer.
        retry_policy: Backoff schedule for transient failures.
        timeout_s: Per-request HTTP timeout.
        client: Inject a pre-built ``httpx.Client`` for testing.
    """

    def __init__(
        self,
        endpoint: str,
        *,
        auth_token: str | None = None,
        batch_size: int = 100,
        batch_max_wait_s: float = 1.0,
        retry_policy: RetryPolicy = RetryPolicy(),
        timeout_s: float = 10.0,
        client: httpx.Client | None = None,
    ) -> None:
        self._endpoint = endpoint
        self._auth_token = auth_token
        self._batch_size = batch_size
        self._batch_max_wait_s = batch_max_wait_s
        self._retry = retry_policy
        self._client = client or httpx.Client(timeout=timeout_s)
        self._owns_client = client is None

        self._buffer: list[Span] = []
        self._lock = threading.Lock()
        self._has_work = threading.Event()
        self._shutdown = threading.Event()
        self._dropped = 0
        self._first_enqueue_time: float | None = None

        self._worker = threading.Thread(
            target=self._flush_loop,
            name="agent-capture-http",
            daemon=True,
        )
        self._worker.start()

    @property
    def dropped_count(self) -> int:
        with self._lock:
            return self._dropped

    # ---- SpanExporter -----------------------------------------------------

    def export(self, span: Span) -> None:
        if self._shutdown.is_set():
            return
        with self._lock:
            self._buffer.append(span)
            if self._first_enqueue_time is None:
                self._first_enqueue_time = time.monotonic()
            full = len(self._buffer) >= self._batch_size
        self._has_work.set()
        if full:
            # Optimistic flush hint; the worker handles the actual send so
            # the caller never blocks on I/O.
            self._has_work.set()

    def flush(self, timeout: float = 5.0) -> None:
        """Synchronously send the current buffer. Does not block the worker.

        Lets a caller ship within a request window on platforms that throttle
        background-thread CPU between requests (e.g. Cloud Run). Never raises.
        """
        with self._lock:
            batch = self._buffer
            self._buffer = []
            self._first_enqueue_time = None
        if batch:
            self._send_with_retry(batch)

    def shutdown(self, timeout: float = 5.0) -> None:
        if self._shutdown.is_set():
            return
        self._shutdown.set()
        self._has_work.set()
        self._worker.join(timeout=timeout)
        # Final drain in case the worker missed something.
        with self._lock:
            remaining = self._buffer
            self._buffer = []
        if remaining:
            self._send_with_retry(remaining)
        if self._owns_client:
            with contextlib.suppress(Exception):
                self._client.close()

    # ---- internals --------------------------------------------------------

    def _flush_loop(self) -> None:
        while True:
            self._has_work.wait(timeout=self._batch_max_wait_s)
            self._has_work.clear()
            if self._shutdown.is_set() and not self._buffer:
                return
            self._maybe_flush()
            if self._shutdown.is_set() and not self._buffer:
                return

    def _maybe_flush(self) -> None:
        with self._lock:
            if not self._buffer:
                self._first_enqueue_time = None
                return
            age = time.monotonic() - self._first_enqueue_time if self._first_enqueue_time is not None else 0.0
            should_flush = (
                len(self._buffer) >= self._batch_size or age >= self._batch_max_wait_s or self._shutdown.is_set()
            )
            if not should_flush:
                return
            batch = self._buffer
            self._buffer = []
            self._first_enqueue_time = None
        self._send_with_retry(batch)

    def _send_with_retry(self, batch: list[Span]) -> None:
        if not batch:
            return
        try:
            payload = self._serialize(batch)
        except Exception as exc:
            log_error(
                ErrorCode.AC402,
                "HTTPExporter: serialization failed; dropping %d spans: %s",
                len(batch),
                exc,
            )
            with self._lock:
                self._dropped += len(batch)
            return

        def attempt() -> None:
            resp = self._client.post(
                self._endpoint,
                content=payload,
                headers=self._headers(),
            )
            # Permanent failures (4xx) are non-retryable. Raise an
            # exception that the retry predicate identifies as permanent.
            if 400 <= resp.status_code < 500:
                raise _PermanentHTTPError(resp.status_code, resp.text[:512])
            resp.raise_for_status()

        try:
            with_retry(
                attempt,
                policy=self._retry,
                retryable=lambda exc: not isinstance(exc, _PermanentHTTPError),
            )
        except _PermanentHTTPError as exc:
            log_error(
                ErrorCode.AC403,
                "HTTPExporter: ledger rejected batch of %d spans with HTTP %d: %s",
                len(batch),
                exc.status,
                exc.body,
                exc_info=False,
            )
            with self._lock:
                self._dropped += len(batch)
        except Exception as exc:
            log_error(
                ErrorCode.AC404,
                "HTTPExporter: dropping %d spans after retries: %s",
                len(batch),
                exc,
            )
            with self._lock:
                self._dropped += len(batch)

    def _serialize(self, batch: list[Span]) -> bytes:
        # Use Pydantic's per-span JSON dump and stitch by hand — avoids
        # an extra dict round-trip and keeps the canonical bytes intact.
        body: dict[str, Any] = {
            "spans": [json.loads(s.model_dump_json(exclude_none=False)) for s in batch],
        }
        return json.dumps(body, separators=(",", ":")).encode("utf-8")

    def _headers(self) -> dict[str, str]:
        h = {"content-type": "application/json"}
        if self._auth_token:
            h["authorization"] = f"Bearer {self._auth_token}"
        return h


class _PermanentHTTPError(Exception):
    """Raised on 4xx so the retry helper skips further attempts."""

    def __init__(self, status: int, body: str) -> None:
        super().__init__(f"HTTP {status}: {body}")
        self.status = status
        self.body = body
